### PR TITLE
add example with traffic lights

### PR DESCRIPTION
**Pull request by AlirezaMorsali**

### CODE REVIEW

**Review by Ruishenl — APPROVED** [review]

File: invertedai/cosimulation.py [hunk 1]
```diff
@@ -111,7 +111,8 @@ def ego_agent_mask(self, value):
     @property
     def ego_states(self):
         """
-        The predicted states for ego_agents only.
+        Returns the current states of ego agents in order.
```
Comment (adscib) [inline invertedai/cosimulation.py:114]: Why did you change "predicted" to "current"? I used "predicted" to emphasize that this is one step ahead of the local simulator.

File: invertedai/common.py [hunk 2]
```diff
@@ -73,6 +73,10 @@ def __init__(self, encoded_image: List[int]):
         self._encoded_image = encoded_image
 
     def decode(self):
+        """
+        Decode and return the image .
```
Comment (adscib) [inline invertedai/common.py:77]: What's up with `.`? Is that just an unneeded space?

File: invertedai/common.py [hunk 3]
```diff
@@ -124,9 +128,13 @@ class AgentAttributes:
 
     length: float  #: Longitudinal extent of the agent, in meters.
     width: float  #: Lateral extent of the agent, in meters.
-    rear_axis_offset: float  #: Distance from the agent's center to its rear axis in meters. Determines motion constraints.
+    #: Distance from the agent's center to its rear axis in meters. Determines motion constraints.
```
Comment (adscib) [inline invertedai/common.py:131]: Did you check that this renders correctly in sphinx?

File: invertedai/common.py [hunk 4]
```diff
@@ -124,9 +128,13 @@ class AgentAttributes:
...
+    rear_axis_offset: float
 
     def tolist(self):
+        """
+        Convert AgentAttributes to a flattened list of agent attributes in this order: [length, width, rear_axis_offset]
```
Comment (adscib) [inline invertedai/common.py:136]: Break line.

File: invertedai/common.py [hunk 5]
```diff
@@ -179,12 +194,16 @@ class StaticMapActor:
...
-    orientation: float  #: Natural direction of traffic going through the stop line, in radians like in :class:`AgentState`.
+    #: Natural direction of traffic going through the stop line, in radians like in :class:`AgentState`.
+    orientation: float
     length: float  #: Size of the stop line, in meters, along its `orientation`.
     width: float  #: Size of the stop line, in meters, across its `orientation`.
 
     @classmethod
     def fromdict(cls, d):
+        """
+        Build StaticMapActor from a dictionary with keys: `actor_id`, `agent_type`, `orientation`, `length`, `width`, `x`, `y`
```
Comment (adscib) [inline invertedai/common.py:205]: Break line.

File: examples/npc_only_with_traffic_lights_REST.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+agent_attributes = response.agent_attributes
+frames = []
+pbar = tqdm(range(50))
+for i in pbar:
+    traffic_light_counter += 1
+    response = iai.drive(
+        agent_attributes=agent_attributes,
+        agent_states=response.agent_states,
+        recurrent_states=response.recurrent_states,
+        traffic_lights_states=df[traffic_light_counter%600].to_dict(),
```
Comment (adscib) [inline examples/npc_only_with_traffic_lights_REST.py:52]: If this is dataframe length, use that instead of the magic number.

File: examples/npc_only_with_traffic_lights_REST.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
+from PIL import Image as PImage
```
Comment (adscib) [inline examples/npc_only_with_traffic_lights_REST.py:1]: Add something to the README to inform people what this example does.

**Review by adscib — CHANGES_REQUESTED** [review]
Looks good. My comments are mostly related to the other changes.

File: invertedai/common.py (hunk 3 above)
Comment (AlirezaMorsali) [reply invertedai/common.py:131]: Yes, it does.

File: examples/npc_only_with_traffic_lights_REST.py (hunk 7 above)
Comment (AlirezaMorsali) [reply examples/npc_only_with_traffic_lights_REST.py:1]: Added the following:
## Controlling the traffic-lights
Inverted AI API allows for controlling the traffic light states. 
We provided a [sample code](https://colab.research.google.com/github/inverted-ai/invertedai-drive/blob/develop/examples/npc_only_with_traffic_lights_REST.py)
for doing so at a traffic-light controlled intersection.

File: examples/npc_only_with_traffic_lights_REST.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+agent_attributes = response.agent_attributes
+frames = []
+pbar = tqdm(range(50))
+for i in pbar:
+    traffic_light_counter += 1
+    response = iai.drive(
+        agent_attributes=agent_attributes,
+        agent_states=response.agent_states,
+        recurrent_states=response.recurrent_states,
+        traffic_lights_states=df[traffic_light_counter%len(df.columns)].to_dict(),
```
Comment (adscib) [inline examples/npc_only_with_traffic_lights_REST.py:52]: Doesn't `len(df.columns)` return the number of columns?

Comment (AlirezaMorsali) [reply examples/npc_only_with_traffic_lights_REST.py:61]: Yes, the columns represent time. Since we are not using this data set I didn't change the format.

**Comment by AlirezaMorsali** [comment]
@adscib @Ruishenl 
I included carla integration with traffic light control in this PR as it was already open.

File: examples/carla/carla_simulator.py [hunk 9]
```diff
@@ -183,10 +185,41 @@ def __init__(
...
-        # pick spawn points for NPCs
+        # Get Traffic-light IDS
+        self.static_actors = static_actors
+        self.tl_objs = list(world.get_actors().filter('traffic.traffic_light*'))
+        self.traffic_lights = {}
+        if static_actors:
+            traffic_lights_obj = list(world.get_actors().filter('traffic.traffic_light*'))
+            for tl in static_actors:
+                if tl.agent_type == "traffic-light-actor":
+                    for tlo in (traffic_lights_obj):
```
Comment (Ruishenl) [inline examples/carla/carla_simulator.py:196]: redundant parentheses

File: invertedai/common.py [hunk 10]
```diff
@@ -179,12 +195,18 @@ class StaticMapActor:
...
+    orientation: float
+    length: Optional[float]  #: Size of the stop line, in meters, along its `orientation`.
+    width: Optional[float]  #: Size of the stop line, in meters, across its `orientation`.
+    dependant: Optional[List[int]]  # : List of ID's of other actors that are dependant to this actor.
 
     @classmethod
     def fromdict(cls, d):
+        """
+        Build StaticMapActor from a dictionary
+        with keys: `actor_id`, `agent_type`, `orientation`, `length`, `width`, `x`, `y`
```
Comment (Ruishenl) [inline invertedai/common.py:208]: maybe add "dependant: as well in the docstring?

**Review by Ruishenl — CHANGES_REQUESTED** [review]
Just some minor comments, and also the default location in CarlaSimulationConfig should be lower case carla not uppser case now

**Review by Ruishenl — APPROVED** [review]
